common/relative_clock.hpp

```diff
@@ -0,0 +1,59 @@
+#pragma once
+#include <chrono>
+#include <ratio>
+#include "phonebook.hpp"
+
+namespace ILLIXR {
+
+//using time_type [[deprecated("Use `RelativeClock::time_point`")]] = std::chrono::steady_clock::time_point;
+
+/**
+ * @brief Relative clock for all of ILLIXR
+ *
+ * Please use this instead of std::chrono clocks; this way, you can fake real time without changing your code.
+ *
+ * It also eliminates the class of bugs relating to using absolute time instead of time-since-start.
+ *
+ * Unfortunately this can't satisfy [Clock][1]
+ * because it needs to have data (namely _m_start) shared across link-time boundaries. There's no
+ * clean way to do this with static variables, so instead I use instance variables and Phonebook.
+ *
+ * [1]: https://en.cppreference.com/w/cpp/named_req/Clock
+ */
+class RelativeClock : public phonebook::service {
+public:
+	using rep = long;
+	using period = std::nano;
+	using duration = std::chrono::duration<rep, period>;
+	using time_point = std::chrono::time_point<RelativeClock>;
```

Comment:
We actually never need to use `RelativeClock` as a `Clock`. We're using `RelativeClock` as a `RelativeClock`.

Here's an example of using `Clock` (aka `CppClock`) and using `RelativeClock`.

```
template<typename Clock>
void useCppClock() {
    Clock::time_point point = Clock::now():
}

void useOurClock(const RelativeClock& clock) {
    RelativeClock::time_point = clock.now():
}

int main() {
    useCppClock<std::chrono::steady_clock>();
    // unfortunately, useCppClock<RelativeClock>() won't work because RelativeClock needs instance state.

    RelativeClock clock;
    useOurClock(clock);
    return 0;
}
```

Reply:
I'm sorry, I still don't understand. My question stems from the [time_point documentation](https://en.cppreference.com/w/cpp/chrono/time_point). The first template parameter is `Clock`, and the documentation says that
> Clock must meet the requirements for Clock

which `RelativeClock` doesn't. So I'm just wondering how this whole thing is working.

Reply:
Ok, I see what you are saying now.

Since the code is templated, the compiler doesn't care that we don't have `RelativeClock::now()` until we invoke code that uses it (like `std::this_thread::sleep_until`). We stick to just doing operations like `time_point + duration` which don't ever call `now()`.

I actually didn't notice that clause in the documentation. With this knowledge, I consider this code bad practice. In C++20, they will be able to enforce type constraints more rigorously, and this would not compile.

The solution is to write our own `time_point` class instead of using `std::chrono::time_point`. It just needs to support arithmetic with `duration`. I will write that soon.

Reply:
Created our own time_point in the latest commit.